lib/cagen.py

```diff
@@ -309,8 +317,50 @@ def _get_hostname():
     except socket.error:
         return None
 
-def _write_file(path, contents):
+def _write_file(path, contents, mode=0644):
     """Utility function for writing to a file"""
-    f = open(path, 'w')
-    f.write(contents)
-    f.close()
+    tmp_file = tempfile.NamedTemporaryFile(dir=os.path.dirname(path), delete=False)
+    tmp_file.write(contents)
+    tmp_file.flush()
+    _safe_move(tmp_file, path, mode)
+
+def _safe_move(new_file, target_path, mode=0644, uid=0, gid=0):
+    """
+    Move 'new_file' (file, NamedTemporaryFile, or path) to 'target_path', backing up 'target_path' to 'target_path.old'
+    if it already exists. Maintain the permissions and ownership of 'target_path' or default to 0644 owned by root:root
+    if 'target_path' does not exist. If the contents of 'new_file' are the same as the 'target_path', do nothing.
+    """
+    if isinstance(new_file, str):
+        new_path = new_file
+        with open(new_path, 'r') as new_file:
+            contents = new_file.read()
+    elif isinstance(new_file, file) or hasattr(new_file, 'file'): # NamedTemporaryFiles have a 'file' attribute
+        new_path = new_file.name
+        new_file.seek(0)
+        contents = new_file.read()
+    else:
+        raise TypeError('Expected string, file, or NamedTemporaryFile instance')
+
+    os.chmod(new_path, mode)
+    os.chown(new_path, uid, gid)
+
+    try:
+        with open(target_path, 'r') as old_file:
+            old_contents = old_file.read()
+        if contents.strip() == old_contents.strip():
+            os.remove(new_path)
+            return
+
+        stat = os.stat(target_path)
+        mode = stat.st_mode
+        uid = stat.st_uid
+        gid = stat.st_gid
+        os.rename(target_path, target_path + '.old')
+    except IOError, exc:
```

Comment:
File not found raises `OSError`, not `IOError`. Also see above quibble about exception syntax.

Reply:
Opening a non-existent file actually does raise IOError. Should I catch `IOError`, and `OSError`?

Reply:
Interesting! The Python manual is a little crappy on the subject. In this case, you should change the except statement in `usercert()` above.

To be honest, I never remember which problems are IOError and which ones are OSError, so I always catch their base class, EnvironmentError.

Reply:
[This one](https://github.com/opensciencegrid/osg-ca-generator/pull/7/files#diff-aa56a9951831a6f2695c73c36d214137R134)? `os.makedirs` and `os.chown` raise `OSError` if the folder already exists or if the folder doesn't exist, respectively.

Reply:
So for nonexistent files it's IOError and for nonexistent directories it's OSError? This is terrible.

Apparently in Python 3, IOError = OSError = EnvironmentError, so the Python devs agree with me.

Reply:
I think it depends on the function:

* `open()` on a non-existent file throws IOError
* `os.chown()` on a non-existent file OR directory throws OSError

Regardless, I can change them all to EnvironmentError and be done with it

Reply:
If you're sure it works then you don't need to, but it would make me more comfortable if you did.